File: src/main/java/com/codebutler/farebot/card/cepas/CEPASProtocol.java
/*
 * CEPASProtocol.java
 *
 * This file is part of FareBot.
 * Learn more at: https://codebutler.github.io/farebot/
 *
 * Copyright (C) 2011-2012, 2015 Eric Butler <eric@codebutler.com>
 * Copyright (C) 2011 Sean Cross <sean@chumby.com>
 * Copyright (C) 2012 tbonang <bonang@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.codebutler.farebot.card.cepas;

import android.nfc.tech.IsoDep;
import android.support.annotation.NonNull;
import android.util.Log;

import com.codebutler.farebot.card.cepas.raw.RawCEPASHistory;
import com.codebutler.farebot.card.cepas.raw.RawCEPASPurse;

import java.io.ByteArrayOutputStream;
import java.io.IOException;

class CEPASProtocol {

    private static final String TAG = "CEPASProtocol";

    private static final byte[] CEPAS_SELECT_FILE_COMMAND = new byte[]{
            (byte) 0x00, (byte) 0xA4, (byte) 0x00, (byte) 0x00,
            (byte) 0x02, (byte) 0x40, (byte) 0x00};

    /* Status codes */
    private static final byte OPERATION_OK = (byte) 0x00;
    private static final byte PERMISSION_DENIED = (byte) 0x9D;

    @NonNull private final IsoDep mTagTech;

    CEPASProtocol(@NonNull IsoDep tagTech) {
        mTagTech = tagTech;
    }

    RawCEPASPurse getPurse(int purseId) throws IOException {
        try {
            sendSelectFile();
            byte[] purseBuff = sendRequest((byte) 0x32, (byte) (purseId), (byte) 0, (byte) 0, new byte[]{(byte) 0});
            if (purseBuff != null) {
                return RawCEPASPurse.create(purseId, purseBuff);
            } else {
                return RawCEPASPurse.create(purseId, "No purse found");
            }
        } catch (CEPASException ex) {
            Log.w(TAG, "Error reading purse " + purseId, ex);
            return RawCEPASPurse.create(purseId, ex.getMessage());
        }
    }

    @NonNull
    RawCEPASHistory getHistory(int purseId, int recordCount) throws IOException {
        try {
            byte[] fullHistoryBuff = null;
            byte[] historyBuff = sendRequest((byte) 0x32, (byte) (purseId), (byte) 0, (byte) 1,
                    new byte[]{(byte) 0, (byte) (recordCount <= 15 ? recordCount * 16 : 15 * 16)});

            if (historyBuff != null) {
                if (recordCount > 15) {
                    byte[] historyBuff2 = null;
                    try {
                        historyBuff2 = sendRequest((byte) 0x32, (byte) (purseId), (byte) 0, (byte) 1,
                                new byte[]{(byte) 0x0F, (byte) ((recordCount - 15) * 16)});
                    } catch (CEPASException ex) {
                        Log.w(TAG, "Error reading 2nd purse history " + purseId, ex);
                    }
                    fullHistoryBuff = new byte[historyBuff.length + (historyBuff2 != null ? historyBuff2.length : 0)];

                    System.arraycopy(historyBuff, 0, fullHistoryBuff, 0, historyBuff.length);
                    if (historyBuff2 != null) {
                        System.arraycopy(historyBuff2, 0, fullHistoryBuff, historyBuff.length, historyBuff2.length);
                    }
                } else {
                    fullHistoryBuff = historyBuff;
                }
            }

            if (fullHistoryBuff != null) {
                return RawCEPASHistory.create(purseId, fullHistoryBuff);
            } else {
                return RawCEPASHistory.create(purseId, "No history found");
            }
        } catch (CEPASException ex) {
            Log.w(TAG, "Error reading purse history " + purseId, ex);
            return RawCEPASHistory.create(purseId, ex.getMessage());
        }
    }

    private byte[] sendSelectFile() throws IOException {
        return mTagTech.transceive(CEPAS_SELECT_FILE_COMMAND);
    }

    private byte[] sendRequest(byte command, byte p1, byte p2, byte lc, byte[] parameters) throws CEPASException,
            IOException {
        ByteArrayOutputStream output = new ByteArrayOutputStream();

        byte[] recvBuffer = mTagTech.transceive(wrapMessage(command, p1, p2, lc, parameters));

        if (recvBuffer[recvBuffer.length - 2] != (byte) 0x90) {
            if (recvBuffer[recvBuffer.length - 2] == 0x6b) {
                throw new CEPASException("File " + p1 + " was an invalid file.");

            } else if (recvBuffer[recvBuffer.length - 2] == 0x67) {
                throw new CEPASException("Got invalid file size response.");
            }

            throw new CEPASException("Got generic invalid response: "
                    + Integer.toHexString(((int) recvBuffer[recvBuffer.length - 2]) & 0xff));
        }

        output.write(recvBuffer, 0, recvBuffer.length - 2);

        byte status = recvBuffer[recvBuffer.length - 1];
        if (status == OPERATION_OK) {
            return output.toByteArray();
        } else if (status == PERMISSION_DENIED) {
            throw new CEPASException("Permission denied");
        } else {
            throw new CEPASException("Unknown status code: " + Integer.toHexString(status & 0xFF));
        }
    }

    private byte[] wrapMessage(byte command, byte p1, byte p2, byte lc, byte[] parameters) throws IOException {
        ByteArrayOutputStream stream = new ByteArrayOutputStream();

        stream.write((byte) 0x90); // CLA
        stream.write(command);     // INS
        stream.write(p1);          // P1
        stream.write(p2);          // P2
        stream.write(lc);          // Lc

        // Write Lc and data fields
        if (parameters != null) {
            stream.write(parameters); // Data field
        }

        return stream.toByteArray();
    }
}

File: farebot-card-cepas/src/commonMain/kotlin/com/codebutler/farebot/card/cepas/CEPASProtocol.kt
/*
 * CEPASProtocol.kt
 *
 * This file is part of FareBot.
 * Learn more at: https://codebutler.github.io/farebot/
 *
 * Copyright (C) 2011-2012, 2015-2016 Eric Butler <eric@codebutler.com>
 * Copyright (C) 2011 Sean Cross <sean@chumby.com>
 * Copyright (C) 2012 tbonang <bonang@gmail.com>
 * Copyright (C) 2016 Michael Farrell <micolous+git@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.codebutler.farebot.card.cepas

import com.codebutler.farebot.card.cepas.raw.RawCEPASHistory
import com.codebutler.farebot.card.cepas.raw.RawCEPASPurse
import com.codebutler.farebot.card.nfc.CardTransceiver

internal class CEPASProtocol(private val mTransceiver: CardTransceiver) {

    fun getPurse(purseId: Int): RawCEPASPurse {
        try {
            sendSelectFile()
            val purseBuff = sendRequest(0x32.toByte(), purseId.toByte(), 0.toByte(), 0.toByte(), byteArrayOf(0.toByte()))
            return if (purseBuff != null) {
                RawCEPASPurse.create(purseId, purseBuff)
            } else {
                RawCEPASPurse.create(purseId, "No purse found")
            }
        } catch (ex: CEPASException) {
            return RawCEPASPurse.create(purseId, ex.message ?: "Unknown error")
        }
    }

    fun getHistory(purseId: Int, recordCount: Int): RawCEPASHistory {
        try {
            var fullHistoryBuff: ByteArray? = null
            val historyBuff = sendRequest(0x32.toByte(), purseId.toByte(), 0.toByte(), 1.toByte(),
                byteArrayOf(0.toByte(), (if (recordCount <= 15) recordCount * 16 else 15 * 16).toByte()))

            if (historyBuff != null) {
                if (recordCount > 15) {
                    var historyBuff2: ByteArray? = null
                    try {
                        historyBuff2 = sendRequest(0x32.toByte(), purseId.toByte(), 0.toByte(), 1.toByte(),
                            byteArrayOf(0x0F.toByte(), ((recordCount - 15) * 16).toByte()))
                    } catch (ex: CEPASException) {
                        // Error reading 2nd purse history
                    }
                    fullHistoryBuff = ByteArray(historyBuff.size + (historyBuff2?.size ?: 0))

                    historyBuff.copyInto(fullHistoryBuff, 0)
                    if (historyBuff2 != null) {
                        historyBuff2.copyInto(fullHistoryBuff, historyBuff.size)
                    }
                } else {
                    fullHistoryBuff = historyBuff
                }
            }

            return if (fullHistoryBuff != null) {
                RawCEPASHistory.create(purseId, fullHistoryBuff)
            } else {
                RawCEPASHistory.create(purseId, "No history found")
            }
        } catch (ex: CEPASException) {
            return RawCEPASHistory.create(purseId, ex.message ?: "Unknown error")
        }
    }

    private fun sendSelectFile(): ByteArray {
        return mTransceiver.transceive(CEPAS_SELECT_FILE_COMMAND)
    }

    @Throws(CEPASException::class)
    private fun sendRequest(command: Byte, p1: Byte, p2: Byte, lc: Byte, parameters: ByteArray): ByteArray? {
        val recvBuffer = mTransceiver.transceive(wrapMessage(command, p1, p2, lc, parameters))

        if (recvBuffer[recvBuffer.size - 2] != 0x90.toByte()) {
            if (recvBuffer[recvBuffer.size - 2] == 0x6b.toByte()) {
                throw CEPASException("File $p1 was an invalid file.")
            } else if (recvBuffer[recvBuffer.size - 2] == 0x67.toByte()) {
                throw CEPASException("Got invalid file size response.")
            }

            throw CEPASException("Got generic invalid response: "
                    + (recvBuffer[recvBuffer.size - 2].toInt() and 0xff).toString(16))
        }

        val output = recvBuffer.copyOfRange(0, recvBuffer.size - 2)

        val status = recvBuffer[recvBuffer.size - 1]
        return when (status) {
            OPERATION_OK -> output
            PERMISSION_DENIED -> throw CEPASException("Permission denied")
            else -> throw CEPASException("Unknown status code: " + (status.toInt() and 0xFF).toString(16))
        }
    }

    private fun wrapMessage(command: Byte, p1: Byte, p2: Byte, lc: Byte, parameters: ByteArray?): ByteArray {
        val paramSize = parameters?.size ?: 0
        val result = ByteArray(5 + paramSize)
        var offset = 0

        result[offset++] = 0x90.toByte() // CLA
        result[offset++] = command     // INS
        result[offset++] = p1          // P1
        result[offset++] = p2          // P2
        result[offset++] = lc          // Lc

        if (parameters != null) {
            parameters.copyInto(result, offset)
        }

        return result
    }

    companion object {
        private val CEPAS_SELECT_FILE_COMMAND = byteArrayOf(
            0x00.toByte(), 0xA4.toByte(), 0x00.toByte(), 0x00.toByte(),
            0x02.toByte(), 0x40.toByte(), 0x00.toByte()
        )

        /* Status codes */
        private const val OPERATION_OK: Byte = 0x00.toByte()
        private val PERMISSION_DENIED: Byte = 0x9D.toByte()
    }
}
